libnfs4acl/acl_nfs4_inheritance.c

```diff
@@ -0,0 +1,563 @@
+		group_allow |= NFS4_ACE_READ_DATA;
+	}
+	if (mode & S_IWGRP) {
+		group_allow |= NFS4_ACE_WRITE_DATA;
+	}
+	if (mode & S_IXGRP) {
+		group_allow |= NFS4_ACE_EXECUTE;
+	}
+	if (mode & S_IROTH) {
+		everyone_allow |= NFS4_ACE_READ_DATA;
+	}
+	if (mode & S_IWOTH) {
+		everyone_allow |= NFS4_ACE_WRITE_DATA;
+	}
+	if (mode & S_IXOTH) {
+		everyone_allow |= NFS4_ACE_EXECUTE;
+	}
+	user_deny = ((group_allow | everyone_allow) & ~user_allow);
+	group_deny = everyone_allow & ~group_allow;
+	user_allow_first = group_deny & ~user_deny;
+	if (!skip_mode) {
+		if (user_allow_first != 0) {
+			new_ace = nfs4_new_ace(
+				is_dir, NFS4_ACE_ACCESS_ALLOWED_ACE_TYPE,
+				0, user_allow_first, NFS4_ACL_WHO_OWNER, NULL
+			);
+			if (new_ace == NULL) {
+				return false;
+			}
+			ret = nfs4_append_ace(aclp, new_ace);
+			if (ret != 0) {
+				return false;
+			}
+		}
+		if (user_deny != 0) {
+			new_ace = nfs4_new_ace(
+				is_dir, NFS4_ACE_ACCESS_DENIED_ACE_TYPE,
+				0, user_deny, NFS4_ACL_WHO_OWNER, NULL
+			);
+			if (new_ace == NULL) {
+				return false;
+			}
+			ret = nfs4_append_ace(aclp, new_ace);
+			if (ret != 0) {
+				return false;
+			}
+		}
+		if (group_deny != 0) {
+			new_ace = nfs4_new_ace(
+				is_dir, NFS4_ACE_ACCESS_DENIED_ACE_TYPE,
+				(NFS4_ACE_GROUP | NFS4_ACE_IDENTIFIER_GROUP),
+				group_deny, NFS4_ACL_WHO_GROUP, NULL
+			);
+			if (new_ace == NULL) {
+				return false;
+			}
+			ret = nfs4_append_ace(aclp, new_ace);
+			if (ret != 0) {
+				return false;
+			}
+		}
+	}
+	if (parent_aclp != NULL) {
+		ok = acl_nfs4_inherit_entries(parent_aclp, aclp, is_dir);
+		if (!ok) {
+			return false;
+		}
+	}
+	if (!skip_mode) {
+		new_ace = nfs4_new_ace(
+			is_dir, NFS4_ACE_ACCESS_ALLOWED_ACE_TYPE,
+			0, user_allow, NFS4_ACL_WHO_OWNER, NULL
+		);
+		if (new_ace == NULL) {
+			return false;
+		}
+		ret = nfs4_append_ace(aclp, new_ace);
+		if (ret != 0) {
+			return false;
+		}
+		new_ace = nfs4_new_ace(
+			is_dir, NFS4_ACE_ACCESS_ALLOWED_ACE_TYPE,
+			NFS4_ACE_IDENTIFIER_GROUP,
+			group_allow, NFS4_ACL_WHO_GROUP, NULL
+		);
+		if (new_ace == NULL) {
+			return false;
+		}
+		ret = nfs4_append_ace(aclp, new_ace);
+		if (ret != 0) {
+			return false;
+		}
+		new_ace = nfs4_new_ace(
+			is_dir, NFS4_ACE_ACCESS_ALLOWED_ACE_TYPE,
+			0, everyone_allow, NFS4_ACL_WHO_EVERYONE, NULL
+		);
+		if (new_ace == NULL) {
+			return false;
+		}
+		ret = nfs4_append_ace(aclp, new_ace);
+		if (ret != 0) {
+			return false;
+		}
+	}
+	return true;
+}
+
+
+/*
+ * Evaluate owner@, group@, everyone@ entries and convert into a POSIX mode.
+ */
+bool acl_nfs4_sync_mode_from_acl(mode_t *_mode, struct nfs4_acl *aclp)
+{
+	mode_t old_mode = *_mode, mode = 0, deny_mode = 0;
+	struct nfs4_ace *ace = NULL;
+
+	for (ace = nfs4_get_first_ace(aclp); ace != NULL;
+	     ace = nfs4_get_next_ace(&ace)) {
+		if (ace->type != NFS4_ACE_ACCESS_ALLOWED_ACE_TYPE &&
+		    ace->type != NFS4_ACE_ACCESS_DENIED_ACE_TYPE) {
+			fprintf(stderr, "Invalid ACE type: %d\n", ace->type);
```

Comment:
Is this a common scenario? If so, could this barf to stderr quite a bit? Maybe you want that to happen? Just curious mostly.

Reply:
This would be quite unexpected since we don't allow setting audit / alarm ACEs.

Reply:
Cool, mostly inquisitive about that so appreciate the explanation.